torch/distributions/von_mises.py

```diff
@@ -124,14 +124,19 @@ class VonMises(Distribution):
     has_rsample = False
 
     def __init__(self, loc, concentration, validate_args=None):
-        self.loc, self.concentration = broadcast_all(loc, concentration)
+        self.dtype = loc.dtype
+        self.loc, self.concentration = broadcast_all(loc.to(torch.double), concentration.to(torch.double))
```

Comment:
It seems a safer to use double versions only in `.sample()`, maybe storing a
```
self._concentration = self.concentration.to(torch.double)
```
for use in `self._proposal_r`.  This way `.log_prob()` can remain single precision, and `.mean` won't magically cast to double.

If we're going to use double stuff in `_proposal_r`, we should change it to a lazy property:
```python
@lazy_property
def _loc(self):
    return self.loc.to(torch.double)

@lazy_property
def _concentration(self):
    return self.concentration.to(torch.double)

@lazy_property
def _proposal_r(self):
    if self._concentration < 1e-5:
        # second order Taylor expansion around kappa = 0
        self._proposal_r = 1 / self._concentration + self._concentration
    else:
        tau = 1 + (1 + 4 * self._concentration**2).sqrt()
        rho = (tau - (2 * tau).sqrt()) / (2 * self._concentration)
        self._proposal_r = (1 + rho**2) / (2 * rho)
```
Then we can use `self._loc` and `self._concentration` in `.sample()`, but continue using `.loc` and `concentration` elsewhere.

Reply:
Actually, concentration can be a tensor, so we'll need a `torch.where(self._concentration < 1e-5, -, -)`